uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java

```diff
@@ -630,6 +690,61 @@ private GroupsTree buildGroupsTree() {
             newGroups.put(g.getLocalKey(), g);
         }
 
+        // Process each group in order to create the groups that will represent the Grouper's
+        // folders
+        // and maintain the consistency regarding the membership.
+        for (String groupKey : newGroups.keySet()) {
+
+            boolean stop = false;
+            boolean isAFolder = false;
+            String currentPath = groupKey;
+            while (!stop) {
+                final String containingFolder = getContainingFolder(currentPath);
```

Comment:
If I'm reading this correctly, current path will be a `:` separated list of paths, and this loops until is has processed all folders.
Would something like `for (String containingFolder : groupKey.split(":")) {` work?

Reply:
This should work. This code come from a Java 1.5 version and I migrate only needed parts...

Reply:
Finally this won't works as the id of the group is the entire path and the containing folder is the entire path, so we need to remove only the last part each time we want the containing folder.

As example if you have this group:
Organizations:org_a:students:class_A
the first containing folder is:
Organizations:org_a:students
the next containing folder is:
Organizations:org_a
and the last containing folder is:
Organizations

So you need to make it in this way, with your solution you will need to rebuild the id path on each values.

Reply:
I understand. :thinking:
My main concern is with `while (!stop) {`, using a flag as a `while` conditional tends to be a code smell. :nose: 
What would be the implications of something like:

```java
var groupKeys = groupKey.split(":");
while (groupKeys.size() > 0) {
  var containingFolder = String.join(":", groupKeys);

  // logic goes here

  groupKeys.removeLast();
}
```

Reply:
I'm agree the entire code of the loop could be rewrited ;)  I Will watch it when I will be back from holidays (I can't before I try to solve this issue https://github.com/uPortal-contrib/uPortal-web-components/issues/43)

Reply:
what do you think about that:
```java

for (String groupKey : newGroups.keySet()) {
            String currentPath = groupKey;

            // The group that represents the grouper folder is created.
            final IEntityGroup folderGroup =
                new EntityTestingGroupImpl(currentPath, IPerson.class);
            folderGroup.setCreatorID("System");
            folderGroup.setName(currentPath);
            folderGroup.setDescription("Grouper Folder");
            newFolders.put(currentPath, folderGroup);

            String containingFolder = this.getContainingFolder(currentPath);
            while (!containingFolder.isEmpty()) {
                // Adds the containing folder as a parent of this current path
                List<String> parentsList = newParents.get(currentPath);
               ...
                childrenList.add(currentPath);

                // The remaining prefix of the group has already been processed.
                if (newFolders.containsKey(containingFolder)) {
                   break; 
                }
                currentPath = containingFolder;

                containingFolder = this.getContainingFolder(currentPath);
            }
        }
```

Reply:
That could work :+1: